src/mjolnir.py

```diff
@@ -0,0 +1,225 @@
+        """
+        super().__init__(charm, "mjolnir")
+        self._charm = charm
+        self._charm_state = charm_state
+        self.framework.observe(charm.on.collect_unit_status, self._on_collect_status)
+
+    @property
+    def _pebble_service(self) -> typing.Any:
+        """Return instance of pebble service.
+
+        Returns:
+            instance of pebble service or none.
+        """
+        return getattr(self._charm, "pebble_service", None)
+
+    def _update_peer_data(self, container: ops.model.Container) -> None:
+        """Update peer data if needed.
+
+        Args:
+            container: Synapse container.
+        """
+        # If there is no secret, we use peer relation data
+        # If there is secret, then we update the secret and add the secret id to peer data
+        peer_relation = self._charm.model.get_relation(PEER_RELATION_NAME)
+        if not peer_relation:
+            # there is no peer relation so nothing to be done
+            return
+
+        if JujuVersion.from_environ().has_secrets and not peer_relation.data[self._charm.app].get(
+            SECRET_ID
+        ):
+            # we can create secrets and the one that we need was not created yet
+            logger.debug("Adding secret")
+            admin_user = self.create_admin_user(container)
+            secret = self._charm.app.add_secret({SECRET_KEY: admin_user.access_token})
+            peer_relation.data[self._charm.app].update({SECRET_ID: secret.id})
+            return
+
+        if not JujuVersion.from_environ().has_secrets and not peer_relation.data[
+            self._charm.app
+        ].get(SECRET_KEY):
+            # we can't create secrets and peer data is empty
+            logger.debug("Updating peer relation data")
+            admin_user = self.create_admin_user(container)
+            peer_relation.data[self._charm.app].update({SECRET_KEY: admin_user.access_token})
+
+    def create_admin_user(self, container: ops.model.Container) -> User:
+        """Create an admin user.
+
+        Args:
+            container: Synapse container.
+
+        Returns:
+            User: admin user that was created.
+        """
+        # The username is random because if the user exists, register_user will try to get the
+        # access_token.
+        # But to do that it needs an admin user and we don't have one yet.
+        # So, to be on the safe side, the user name is randomly generated and if for any reason
+        # there is no access token on peer data/secret, another user will be created.
+        #
+        # Using 16 to create a random value but to  be secure against brute-force attacks, please
+        # check the docs:
+        # https://docs.python.org/3/library/secrets.html#how-many-bytes-should-tokens-use
+        username = token_hex(16)
+        return actions.register_user(container, username, True)
+
+    def _on_collect_status(self, event: ops.CollectStatusEvent) -> None:
```

Comment:
I don't totally agree with the implementation of the collect_status hook here for few reasons:
- collect_status is a hook that can be implemented on multiple observers, and there's a priority logic implemented on ops side to choose the final status, therefore checking Synapse container status shouldn't be managed from here
- It shouldn't implement business logic, so I think that `_update_peer_data` and `enable_mjolnir` shouldn't be called from here, it's only collecting status, so check if service exists, check if there's a room_id, define the status and the message and do `add_status`

Reply:
This definitely is something interesting to be discussed.
I think it could be applied [here](https://github.com/canonical/synapse-operator/pull/30#discussion_r1330005594) as well. What do you think?

Reply:
I'll mark this one as resolved to be discussed in a follow-up PR.